### PR TITLE
Remove 'SourceRange.eliminateFIXME'

### PR DESCRIPTION
**Pull request by kyouko-taiga**
This PR removes the `eliminateFIXME` values that had been used to replace optional source ranges.

Note that we will probably need to determine better source sites for IR instructions during lowering in the future, so as to improve the accuracy of diagnostics and step-debugging.

### CODE REVIEW

File: Sources/Core/AST/Decl/BuiltinDecl.swift [hunk 1]
```diff
@@ -1,6 +1,11 @@
 /// An abstract node representing a built-in declaration.
 public struct BuiltinDecl: Decl {
 
-  public var site: SourceRange { .eliminateFIXME }
+  public let site: SourceRange
+
+  public init() {
+    let f = SourceFile(synthesizedText: "/* built-in declaration */")
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/BuiltinDecl.swift:7]: At lease this could be made a static constant, right?

File: Sources/Core/AST/Decl/TopLevelDeclSet.swift [hunk 2]
```diff
@@ -1,16 +1,17 @@
 /// A set of declarations at the top-level of a source file.
 public struct TopLevelDeclSet: Node, LexicalScope {
 
+  public let site: SourceRange
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/TopLevelDeclSet.swift:4]: It's not obvious how/why this change is necessary to serve the purpose of this PR, which makes me wonder if it's necessary.

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift [hunk 3]
```diff
@@ -10,6 +10,16 @@ public struct SubscriptImplDecl: Decl, LexicalScope {
...
+    /// The node wrappedby this instance.
+    public var node: AnyNodeID {
+      switch self {
+      case .expr(let n):
+        return AnyNodeID(n)
+      case .block(let n):
+        return AnyNodeID(n)
+      }
+    }
+
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/SubscriptImplDecl.swift:22]: It's not obvious to me how this change serves the purpose of the PR, which makes me wonder if it's necessary.

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift [hunk 4]
```diff
@@ -10,6 +10,16 @@ public struct SubscriptImplDecl: Decl, LexicalScope {
     /// A block body.
     case block(NodeID<BraceStmt>)
 
+    /// The node wrappedby this instance.
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/SubscriptImplDecl.swift:13]:
```suggestion
    /// The node wrapped by this instance.
```

File: Sources/FrontEnd/Parse/Parser.swift [hunk 5]
```diff
@@ -30,33 +30,20 @@ public enum Parser {
...
-
-    state.ast[module].addSourceFile(translation)
-    ast = state.ast
-
-    return translation
-  }
-
-  /// Parses an instance of `TopLevelDeclSet`.
-  static func parseSourceFile(in state: inout ParserState) -> NodeID<TopLevelDeclSet> {
+    // Parse the input.
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:38]: This change took a function I could digest at a glance, and turned it into a longer function.  `parseSourceFile` may not have been well-named, but are you sure this is an improvement?

File: Sources/FrontEnd/Parse/Parser.swift [hunk 6]
```diff
@@ -566,8 +567,8 @@ public enum Parser {
     if state.isAtTypeScope && !prologue.isStatic {
       receiver = state.insert(
         synthesized: ParameterDecl(
-          identifier: SourceRepresentable(value: "self", range: .eliminateFIXME),
-          site: .eliminateFIXME))
+          identifier: SourceRepresentable(value: "self", range: head.introducerSite),
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:570]: Seeing stuff like this makes me wonder if the site of ParameterDecl shouldn't be computed from the sites of its stored properties, rather than being stored.

File: Sources/FrontEnd/Parse/Parser.swift [hunk 7]
```diff
@@ -710,9 +711,10 @@ public enum Parser {
     withPrologue prologue: DeclPrologue,
     in state: inout ParserState
   ) throws -> NodeID<InitializerDecl>? {
-    // Parse the parts of the declaration.
-    let parser = (take(nameTokenWithValue: "memberwise").and(take(.`init`)))
-    guard let parts = try parser.parse(&state) else { return nil }
+    // Parse the introducer.
+    guard let a = state.take(nameTokenWithValue: "memberwise") else { return nil }
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:715]: Are there changes in here that don't have to do with the point of this PR?  I don't know how to analyze this one.

File: Sources/FrontEnd/Parse/Parser.swift [hunk 8]
```diff
@@ -886,7 +888,9 @@ public enum Parser {
       if let body = try subscriptImplDeclBody.parse(&state) {
         let impl = try buildSubscriptImplDecl(
           in: &state,
-          withIntroducer: SourceRepresentable(value: .let, range: .eliminateFIXME),
+          introducedBy: SourceRepresentable(
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:891]: The change of label is a good one, but seems irrelevant to the purpose of the PR.  I'd appreciate it if renaming steps could be made separately; ones like this don't even require review.  All in all, this is a very large PR and it would be easier to analyze if it only contained relevant changes.

File: Sources/FrontEnd/Parse/Parser.swift [hunk 9]
```diff
@@ -930,9 +934,18 @@ public enum Parser {
     }
   }
 
+  /// Inserts in `state.ast` a subscript implementation declaration introduced with the given
+  /// introducer and body and returns its ID.
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:938]: “the given x”, where x is not in `code voice`, is always a poor tradeoff in a doc comment.  There are basically two good ways to handle cases like this:

```suggestion
  /// Inserts into `state.ast` a subscript implementation having the given `introducer` and `body` returning its ID.
  ```
  
  ```suggestion
  /// Inserts into `state.ast` a subscript implementation introduced by `intro`,  with body `b`, returning its ID.
  ```

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift [hunk 10]
```diff
@@ -10,6 +10,16 @@ public struct SubscriptImplDecl: Decl, LexicalScope {
...
+    public var node: AnyNodeID {
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/SubscriptImplDecl.swift:14]:
```suggestion
    public var base: AnyNodeID {
```

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift (hunk 4 above)
Comment (dabrahams) [inline Sources/Core/AST/Decl/SubscriptImplDecl.swift:13]: Separately, I'm wondering if `ImplDecl` is not redundant.  Isn't every `Impl` also a `Decl`?

File: Sources/FrontEnd/Parse/Parser.swift [hunk 11]
```diff
@@ -930,9 +934,18 @@ public enum Parser {
...
+  ///
+  /// - Parameters:
+  ///   - introducer: The introducer of the declaration, or `nil` if it is implicit. In that case,
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:941]: It's good to think about how to avoid these clauses when possible.  You could have omitted this description if in the summary fragment you'd said something like “explicitly introduced by `intro`.”

File: Sources/FrontEnd/Parse/Parser.swift [hunk 12]
```diff
@@ -930,9 +934,18 @@ public enum Parser {
...
+  ///     it is synthesized as `let`.
+  ///   - body: The body of the declaration, or `nil` if that body should must be synthesized or
+  ///     if the declaration denotes a trait requirement.
+  /// - Requires: `body` is not `nil` if `introducer` is.
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:945]:
```suggestion
  /// - Requires: if `introducer` is `nil`, body is non-`nil`.
```

File: Sources/FrontEnd/Parse/Parser.swift [hunk 13]
```diff
@@ -1034,7 +1042,10 @@ public enum Parser {
 
     // Retrieve or synthesize the type's memberwise initializer.
     var members = parts.1
-    let memberwiseInit = findOrSynthesizeMemberwiseInitDecl(in: &members, updating: &state)
+    let memberwiseInit = getOrCreateMemberwiseInitializer(
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:1045]: I'm not sure the memoization action of this function is useful information to the reader. What about just naming it `memberwiseInitializer`?

Otherwise, I'd delete the comment!

File: Sources/FrontEnd/Parse/Parser.swift [hunk 14]
```diff
@@ -1051,29 +1062,31 @@ public enum Parser {
 
   /// Returns the first memberwise initializer declaration in `members` or synthesizes an implicit
   /// one, appends it into `members`, and returns it.
-  private static func findOrSynthesizeMemberwiseInitDecl(
-    in members: inout [AnyDeclID],
+  private static func getOrCreateMemberwiseInitializer(
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:1065]: This again feels like a change that could have been separated from the PR.

File: Sources/FrontEnd/Parse/Parser.swift [hunk 15]
```diff
@@ -1051,29 +1062,31 @@ public enum Parser {
...
+    ofDeclStartingAt startIndex: SourceFile.Index,
+    among members: inout [AnyDeclID],
     updating state: inout ParserState
   ) -> NodeID<InitializerDecl> {
     for member in members where member.kind == InitializerDecl.self {
       let m = NodeID<InitializerDecl>(rawValue: member.rawValue)
       if state.ast[m].introducer.value == .memberwiseInit { return m }
     }
 
+    let site = state.emptyRange(at: startIndex)
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:1075]: You can do better.  Maybe

```suggestion
    let startOfTypeDecl = state.emptyRange(at: startIndex)
```

File: Sources/FrontEnd/Parse/Parser.swift [hunk 16]
```diff
@@ -2841,12 +2853,14 @@ public enum Parser {
     (maybe(passingConvention)
       .andCollapsingSoftFailures(expr)
       .map({ (state, tree) -> NodeID<ParameterTypeExpr> in
-        state.insert(
+        let site = state.range(from: tree.0?.site.start ?? state.ast[tree.1].site.start)
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift:2856]: Can we do better?  The site of _what_?  It's not just any old site.

File: Sources/FrontEnd/Parse/ParserState.swift [hunk 17]
```diff
@@ -117,6 +117,11 @@ struct ParserState {
     lexer.sourceCode.range(startIndex ..< currentIndex)
   }
 
+  /// Returns an empty site starting and ending at `index`.
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/ParserState.swift:120]:
```suggestion
  /// Returns a range starting and ending at `index`.
```

It's hard to explain, but the use of the word site here comes off a bit weird.  A site is usally _of_ or _for_ something.  This one is disembodied.  Also "empty site" is not really a thing, English-wise.

File: Tests/ValTests/ASTTests.swift [hunk 18]
```diff
@@ -18,17 +18,19 @@ final class ASTTests: XCTestCase {
     var ast = AST()
 
     // Create a module declarations.
+    let input = SourceFile(synthesizedText: "")
+    let site = input.range(input.text.startIndex ..< input.text.endIndex)
```
Comment (dabrahams) [inline Tests/ValTests/ASTTests.swift:22]: May as well use your new `emptyRange` function, right?
Can't we do better than `site` here?

File: Tests/ValTests/ASTTests.swift [hunk 19]
```diff
@@ -39,17 +41,21 @@ final class ASTTests: XCTestCase {
     var ast = AST()
 
     // Create a module.
+    let input = SourceFile(synthesizedText: "")
+    let site = input.range(input.text.startIndex ..< input.text.endIndex)
```
Comment (dabrahams) [inline Tests/ValTests/ASTTests.swift:45]: ditto to the above comments.  But this is a repeated code pattern; please factor it out!

**Review by dabrahams — COMMENTED** [review]
These are just comments for the first commit in the PR.  More to review.

File: Sources/Core/AST/Decl/TopLevelDeclSet.swift (hunk 2 above)
Comment (kyouko-taiga) [reply Sources/Core/AST/Decl/TopLevelDeclSet.swift:4]: I had to remove the computed property, which was:
```var
public var site: SourceRange { .eliminateFIXME }
```

File: Sources/Core/AST/Decl/BuiltinDecl.swift (hunk 1 above)
Comment (kyouko-taiga) [reply Sources/Core/AST/Decl/BuiltinDecl.swift:7]: Yep.

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift (hunk 3 above)
Comment (kyouko-taiga) [reply Sources/Core/AST/Decl/SubscriptImplDecl.swift:22]: This change eliminates boilerplate in the parser to retrieve the site of a subscript's body.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 6 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:570]: It is not obvious (or even possible) to compute the site of a node from its properties if those are nodes referred by their ID. That's because we need the AST to read the site of an expression represented by `AnyExprID`.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 7 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:715]: The changes are related. They serve to identify the start index of the introducer's site.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 8 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:891]: Noted.

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift (hunk 4 above)
Comment (kyouko-taiga) [reply Sources/Core/AST/Decl/SubscriptImplDecl.swift]: That's a good point. I'll make a separate PR.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 11 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:940]: Agreed. In this particular case, I went back and forth with the docs until I decided the clause was necessary to clarify what happens if `introducer` is `nil`. I felt like something saying "explicitly introduced by `introducer`" was missing the fact that the introducer is synthesized as `let` if kept implicit.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 13 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:1044]: I hesitated. The fact that you're asking the same question leads me to think it's fine to just use `memberwiseInitializer `.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 16 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:2855]: I'll use just `s`

File: Tests/ValTests/ASTTests.swift (hunk 19 above)
Comment (kyouko-taiga) [reply Tests/ValTests/ASTTests.swift:45]: Sure, but factoring out the code pattern is probably not something that should be part of this PR.

File: Tests/ValTests/ASTTests.swift (hunk 18 above)
Comment (kyouko-taiga) [reply Tests/ValTests/ASTTests.swift:22]: `emptyRange` is a method of `ParserState`. It's never used anywhere else because the other synthetic sites we create (for `builtinDecl` and `moduleDecl`) are not empty so I'm note sure we should make it a method of `SourceFile`.

Perhaps it would be more relevant to define a property `SourceFile.wholeRange`.

Comment (dabrahams) [reply Tests/ValTests/ASTTests.swift]: `emptyRange` doesn't make sense to me as a method of `ParserState`, though.  What is an empty range of a parser state?  Could be anything.  `state.source.emptyRange` makes sense.

File: Sources/Core/AST/Decl/SubscriptImplDecl.swift (hunk 3 above)
Comment (dabrahams) [reply Sources/Core/AST/Decl/SubscriptImplDecl.swift]: I eventually realized that, and thought I had deleted the comment.  Sorry!

File: Sources/Core/AST/Decl/TopLevelDeclSet.swift (hunk 2 above)
Comment (dabrahams) [reply Sources/Core/AST/Decl/TopLevelDeclSet.swift:4]: Oh, the github diff didn't make the corresponding removal obvious.

File: Tests/ValTests/ASTTests.swift (hunk 18 above)
Comment (kyouko-taiga) [reply Tests/ValTests/ASTTests.swift]: That's just a helper for a longer expression. The parser state has many helper methods that operate on the lever and/or the input being parsed.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 5 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:38]: Yes. I originally kept `parseSourceFile` only to realize it was requiring too many input/outputs to resembles an actual independent unit of computation.

Comment (kyouko-taiga) [reply Sources/FrontEnd/Parse/Parser.swift:38]: FWIW, `parse(_:into:in:diagnostics:)` is barely longer.

File: Tests/ValTests/ASTTests.swift (hunk 18 above)
Comment (dabrahams) [reply Tests/ValTests/ASTTests.swift]: I don't think that's a good idea, FWIW.  It prevents the parser state from representing a clear abstraction; instead it becomes something like the superposition of its parts.

File: Sources/IR/Operands/Instruction/CondBranchInstruction.swift [hunk 20]
```diff
@@ -14,7 +14,14 @@ public struct CondBranchInstruction: Instruction {
   /// The target of the branch if `condition` is false.
   public let targetIfFalse: Block.ID
 
-  public let range: SourceRange?
+  public let site: SourceRange
+
+  init(condition: Operand, targetIfTrue: Block.ID, targetIfFalse: Block.ID, site: SourceRange) {
```
Comment (dabrahams) [inline Sources/IR/Operands/Instruction/CondBranchInstruction.swift:19]: Curious as to the reason you added this (or omitted it before).

**Review by dabrahams — COMMENTED** [review]
Comments on 2nd commit

**Review by dabrahams — APPROVED** [review]

**Comment by dabrahams** [comment]
**Sorry: Approval was premature; not done reviewing yet**

File: Sources/FrontEnd/TypeChecking/ConstraintSolver.swift [hunk 21]
```diff
@@ -910,7 +910,9 @@ extension TypeChecker {
         else { return .incomparable }
 
         // Rank the candidates.
-        switch (refines(lhs, rhs, scope: scope), refines(rhs, lhs, scope: scope)) {
+        let lRefinesR = refines(lhs, rhs, scope: scope, site: program.ast[n].site)
```
Comment (dabrahams) [inline Sources/FrontEnd/TypeChecking/ConstraintSolver.swift:913]: This call reads a bit weirdly.  `site` is a poor label here (arguably `scope` too); what is its role? 

Maybe something like:
```swift
does(lhs, refine: rhs, in: scope, diagnosingConstraintErrorsAt: program.ast[n].site)
```
?

File: Sources/Core/TypedNode.swift [hunk 22]
```diff
@@ -58,10 +58,12 @@ extension NodeIDProtocol {
 }
 
 extension TypedProgram {
+
```
Comment (dabrahams) [inline Sources/Core/TypedNode.swift:66]: This commit shoulda been a separate PR.

File: Sources/Core/TypedNode.swift [hunk 23]
```diff
@@ -70,7 +70,7 @@ extension TypedNode where ID: ConcreteNodeID {
 
   /// The corresponding AST node.
   private var syntax: ID.Subject {
-    program.ast[NodeID(id)!]
```
Comment (dabrahams) [inline Sources/Core/TypedNode.swift:71]: This commit shoulda been a separate PR.

Can you explain why this change is an improvement?  To me it looks like premature optimization that makes the code harder to read.

File: Sources/Core/TypedNode.swift [hunk 24]
```diff
@@ -24,6 +24,11 @@ public struct TypedNode<ID: NodeIDProtocol>: Hashable {
     self.id = id
   }
 
+  /// The site from which self was parsed.
+  public var site: SourceRange {
```
Comment (dabrahams) [inline Sources/Core/TypedNode.swift:28]: isn't this property automatically reflected by the `@dynamicMemberLookup` stuff?

File: Sources/FrontEnd/Parse/Parser.swift [hunk 25]
```diff
@@ -934,8 +934,7 @@ public enum Parser {
     }
   }
 
-  /// Inserts in `state.ast` a subscript implementation declaration introduced with the given
-  /// introducer and body and returns its ID.
```
Comment (dabrahams) [inline Sources/FrontEnd/Parse/Parser.swift]: Yeah, that phrasing is pretty awkward, sorry.  Let me try again:

```suggestion
/// Inserts a subscript having the given `introducer` and `body` into `state.ast`, returning its ID.
```

File: Sources/Core/AST/Decl/BuiltinDecl.swift [hunk 26]
```diff
@@ -1,13 +1,11 @@
...
 
-  public init() {
-    self.site = Self.source.range(Self.source.text.startIndex ..< Self.source.text.endIndex)
-  }
+  public var site: SourceRange { Self.site }
 
-  /// A synthesized source file to define the site of this type's instances.
-  private static let source = SourceFile(synthesizedText: "/* built-in declaration */")
+  /// A synthesized site for all instances.
+  private static let site = SourceFile(synthesizedText: "/* built-in declaration */").wholeRange
```
Comment (dabrahams) [inline Sources/Core/AST/Decl/BuiltinDecl.swift:9]:
```suggestion
private static let siteConstant = SourceFile(synthesizedText: "/* built-in declaration */").wholeRange
```
While it's possible to have static and non-static variables with the same name, it can be confusing.

**Review by dabrahams — APPROVED** [review]
OK, review complete. Good job; I had no idea this bit would require so much work!

File: Sources/IR/Analysis/DefiniteInitializationPass.swift [hunk 27]
```diff
@@ -118,10 +118,11 @@ public struct DefiniteInitializationPass: TransformPass {
               // at the end of the predecessor.
               if lhs != rhs {
                 // Deinitialize the object at the end of the predecessor.
+                let terminator = module.terminator(
+                  of: Block.ID(function: functionID, address: predecessor))!
                 module.insert(
```
[the remark this answers is not in the document]
Comment (kyouko-taiga) [reply Sources/IR/Analysis/DefiniteInitializationPass.swift:123]: That's a possibility, but note that not all instructions will be created from an AST node. Many will be synthesized from IR passes so the initializer that accepts a SourceRange will always be necessary.

One thing that worries me a little is that it will be difficult to maintain consistent site information once we implement aggressive optimizations that can move instructions around. But that's for future me to address.

File: Sources/IR/Operands/Instruction/StoreInstruction.swift [hunk 28]
```diff
@@ -9,12 +9,12 @@ public struct StoreInstruction: Instruction {
...
 
-  public let range: SourceRange?
+  public let site: SourceRange
 
-  init(_ object: Operand, to target: Operand, range: SourceRange? = nil) {
+  init(_ object: Operand, to target: Operand, site: SourceRange) {
     self.object = object
     self.target = target
-    self.range = range
+    self.site = site
```
[the remark this answers is not in the document]
Comment (kyouko-taiga) [reply Sources/IR/Operands/Instruction/StoreInstruction.swift:17]: Okay, sounds reasonable.

File: Sources/IR/Analysis/ImplicitReturnInsertionPass.swift [hunk 29]
```diff
@@ -17,22 +17,22 @@ public struct ImplicitReturnInsertionPass: TransformPass {
...
+      // Nothing to do if there's a terminator instruction.
+      if last.isTerminator { continue }
+
+      if expectedReturnType == .void {
         // Insert missing return instruction.
         module.insert(
-          ReturnInstruction(),
-          at: module.globalEndIndex(of: Block.ID(function: functionID, address: i.address)))
+          ReturnInstruction(site: last.site),
+          at: module.globalEndIndex(of: Block.ID(function: functionID, address: block.address)))
```
[the remark this answers is not in the document]
Comment (kyouko-taiga) [reply Sources/IR/Analysis/ImplicitReturnInsertionPass.swift:31]: Yes, that's because they are indices of the doubly linked list, which have an address property denoting their stable position.

File: Sources/IR/Operands/Instruction/CondBranchInstruction.swift (hunk 20 above)
Comment (kyouko-taiga) [reply Sources/IR/Operands/Instruction/CondBranchInstruction.swift:19]: It was missing before. I added it for consistency. I should probably have done so in another PR. Sorry.

File: Sources/FrontEnd/TypeChecking/ConstraintSolver.swift (hunk 21 above)
Comment (kyouko-taiga) [reply Sources/FrontEnd/TypeChecking/ConstraintSolver.swift:913]: Labels looks good but is the function name intentional?

File: Sources/Core/TypedNode.swift (hunk 23 above)
Comment (kyouko-taiga) [reply Sources/Core/TypedNode.swift:71]: It's for consistency. I use this initializer every time I know that the conversion will succeed. That makes refactoring simpler if one day we want to change that initializer or how it's used.

File: Sources/Core/TypedNode.swift (hunk 24 above)
Comment (kyouko-taiga) [reply Sources/Core/TypedNode.swift:28]: No, it wasn't! Swift was complaining `AnyExprID` was not conforming to `ConcreteNodeID`. I think it's because `site` is a requirement on `Node`, making it part of the existential API.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 6 above)
Comment (dabrahams) [reply Sources/FrontEnd/Parse/Parser.swift:570]: Yeah, the finding best representation for the AST and balancing that with other factors continues to be a challenge.

File: Sources/IR/Analysis/DefiniteInitializationPass.swift (hunk 27 above)
[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/IR/Analysis/DefiniteInitializationPass.swift:123]: I'm not sure that storing source ranges in instructions will make sense in the long run.  Another possiblity is that the IR data structure maintains a mapping between indices or addresses and source locations.  When synthesizing instructions they adopt the source ranges of their "anchor" instructions, i.e. the ones they are inserted `before` or `after`.

File: Sources/FrontEnd/TypeChecking/ConstraintSolver.swift (hunk 21 above)
Comment (dabrahams) [reply Sources/FrontEnd/TypeChecking/ConstraintSolver.swift]: The base name? I went back and forth about it, and hated most of the things I came up with a little bit.  Probably this one most of all.  I want some standard naming convention for context-requiring functions like this…  How about

```
inFact(lhs, refines: rhs, in: scope, diagnosingConstraintErrorsAt: program.ast[n].site)
```

The need for the context distorts the API a bit because in principle it should be a method of `lhs`.  Could do something like

```
self.constraintSource = program.ast[n].site
lhs.refines(rhs, in: scope, context: &self)
```

🤷 it's hard.  This is the sort of problem `TypedNode` was designed to solve more elegantly.

File: Sources/Core/TypedNode.swift (hunk 23 above)
Comment (dabrahams) [reply Sources/Core/TypedNode.swift:71]: I don't buy it.  `T(x)!` is the canonical form for “the conversion could fail dynamically but I know it will succeed.”  It communicates exactly that to the reader, whereas what you wrote is a head-scratcher.

File: Sources/FrontEnd/Parse/Parser.swift (hunk 11 above)
Comment (dabrahams) [reply Sources/FrontEnd/Parse/Parser.swift:941]: I guess I was assuming the reader would know that information from the language semantics.

File: Sources/Core/TypedNode.swift (hunk 23 above)
Comment (kyouko-taiga) [reply Sources/Core/TypedNode.swift:71]: Okay, you may have a point here, but it remains that I consistently used `NodeID(rawValue:)` everywhere else. So I stand by my point: it will be easier to refactor all those common call sites.

Comment (dabrahams) [reply Sources/Core/TypedNode.swift:71]: Maybe so, but readability is way more important than ease of refactoring.

Comment (kyouko-taiga) [reply Sources/Core/TypedNode.swift:71]: Let's open a separate issue if you want to revisit this initializer. Maybe the solution is to remove it altogether. In any case, that's unrelated to this PR.
